src/lcm/model.py

```diff
@@ -0,0 +1,122 @@
+"""Collection of classes that are used by the user to define the model and grids."""
+
+import dataclasses as dc
+from collections.abc import Callable
+from dataclasses import KW_ONLY, dataclass, field
+
+from lcm.exceptions import ModelInitilizationError, format_messages
+from lcm.grids import Grid
+
+
+@dataclass(frozen=True)
+class Model:
+    """A user model which can be processed into an internal model.
+
+    Attributes:
+        description: Description of the model.
+        n_periods: Number of periods in the model.
+        functions: Dictionary of user provided functions that define the functional
+            relationships between model variables. It must include at least a function
+            called 'utility'.
+        choices: Dictionary of user provided choices.
+        states: Dictionary of user provided states.
+
+    """
+
+    description: str | None = None
+    _: KW_ONLY
+    n_periods: int
+    functions: dict[str, Callable] = field(default_factory=dict)
+    choices: dict[str, Grid] = field(default_factory=dict)
+    states: dict[str, Grid] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        type_errors = _validate_attribute_types(self)
+
+        if type_errors:
+            msg = format_messages(type_errors)
+            raise ModelInitilizationError(msg)
+
+        logical_errors = _validate_logical_consistency(self)
+
+        if logical_errors:
+            msg = format_messages(logical_errors)
+            raise ModelInitilizationError(msg)
+
+    def replace(self, **kwargs) -> "Model":
+        """Replace the attributes of the model.
+
+        Args:
+            **kwargs: Keyword arguments to replace the attributes of the model.
+
+        Returns:
+            A new model with the replaced attributes.
+
+        """
+        return dc.replace(self, **kwargs)
+
+
+def _validate_attribute_types(model: Model) -> list[str]:
```

Comment:
Isn't the code from here on mainly duplicating pydantic functionality? So even if we can't use that everywhere, could we do so for a subset and reduce own code for a stable dependency?

Reply:
Yes, we could do that, good point. I've added it to the to-do list.